### PR TITLE
Added working thanos setup to k8s manifests. Adjusted peer logic.

### PR DESCRIPTION
**Pull request by Bplotka**
Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: cmd/thanos/main.go [hunk 1]
```diff
@@ -161,9 +169,79 @@ func joinCluster(logger log.Logger, typ cluster.PeerType, bindAddr, advertiseAdd
...
+				resolvedPeers = append(resolvedPeers, net.JoinHostPort(ip.String(), port))
+			}
+		}
+
+		// Assume direct address.
+		resolvedPeers = append(resolvedPeers, peer)
+	}
+
+	return resolvedPeers, nil
+}
```
Comment (fabxc) [inline cmd/thanos/main.go]: Quite a few indentions. I think we can reduce that by a few by swapping the conditions and using `continue` etc.

File: cmd/thanos/query.go [hunk 2]
```diff
@@ -36,30 +36,29 @@ func registerQuery(m map[string]setupFunc, app *kingpin.Application, name string
...
-	m[name] = func(logger log.Logger, metrics *prometheus.Registry) (okgroup.Group, error) {
-		peer, err := joinCluster(
+	m[name] = func(logger log.Logger, req *prometheus.Registry) (okgroup.Group, error) {
+		joinConfig, err := createJoinConfig(
 			logger,
-			cluster.PeerTypeQuery,
 			*clusterBindAddr,
-			*clusterAdvertiseAddr,
-			*apiAddr,
+			os.ExpandEnv(*clusterAdvertiseAddr),
```
Comment (fabxc) [inline cmd/thanos/query.go]: Is this for the `POD_IP` envvar? I'd have expected that to be subtituted already by whatever starts the process, no? It's rather awkward semantics to have this be done process-internal to some flags.

File: pkg/cluster/cluster.go [hunk 3]
```diff
@@ -91,20 +91,37 @@ func NewPeer(
...
 	defer p.mtx.RUnlock()
-	p.data[p.Name()] = PeerState{Type: typ, APIAddr: apiAddr}
+
+	p.data[p.Name()] = state
 
 	return p, nil
 }
 
+// SetLabels sets the labels and notifies other members.
+func (p *Peer) SetLabels(labels []storepb.Label, timeout time.Duration) error {
```
Comment (fabxc) [inline pkg/cluster/cluster.go]: Where is this being used?

File: cmd/thanos/query.go (hunk 2 above)
Comment (Bplotka) [reply cmd/thanos/query.go]: Nope it is not substituted - that's the problem ):

File: pkg/cluster/cluster.go (hunk 3 above)
Comment (Bplotka) [reply pkg/cluster/cluster.go]: .. it should be used after querying correct labels. Will add that (:

Comment (fabxc) [reply pkg/cluster/cluster.go]: I'm a bit concerned about the timeout. The purpose of the gossip layer is ultimately that we can distribute state in an eventually consistent way without worrying about it reaching peers etc.

Adding a timeout and error here re-surfaces this concern. What's the appropriate timeout? It failed, now what? Does the sidecar retry indefinitely, or for 10 minutes?

In the end the gossip layer does all this for us. I'd just set the internal label state and call it a day. I'm also not quite sure that `UpdateNode` actually does anything for our custom state. The meta info it refers to is probably the 512 byte slice of Node which we don't use: https://godoc.org/github.com/hashicorp/memberlist#TransmitLimitedQueue.QueueBroadcast

Comment (Bplotka) [reply pkg/cluster/cluster.go]: hm.. yeah. And we need to have some test for that as well. 

Ok let me remove SetLabel from this PR. This PR is starting touching too many things.

**Comment by Bplotka** [comment]
Splitting this PR into smaller bits for easier review

**Pull request closed without merging**